numpy/lib/function_base.py

```diff
@@ -689,16 +699,38 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
                 b = a[keep]
 
         if b.size == 0:
-            bins = 1
+            n_equal_bins = 1
         else:
             # Do not call selectors on empty arrays
-            width = _hist_bin_selectors[bins](b)
+            width = _hist_bin_selectors[bin_name](b)
             if width:
-                bins = int(np.ceil((mx - mn) / width))
+                n_equal_bins = int(np.ceil((mx - mn) / width))
             else:
                 # Width can be zero for some estimators, e.g. FD when
                 # the IQR of the data is zero.
-                bins = 1
+                n_equal_bins = 1
+
+    elif np.ndim(bins) == 0:
+        try:
+            n_equal_bins = operator.index(bins)
+        except TypeError:
+            raise TypeError('If a scalar, `bins` must be an integer')
```

Comment:
Maybe ``` 'If `bins` is a scalar, it must be an integer'```

Same for next error message.

Reply:
Or perhaps `` `bins` must be an integer, if a scalar ``, to try and start all the messages with `bins`?

Reply:
Let me think about it a bit more.

Reply:
Thinking again, this particular error message should be the generic one - see my update,